// msint-identity-aad/lib/index.js
/*!---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *----------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("underscore");
/**
 * The identifier of the Azure Active Directory account provider.
 */
exports.id = "Microsoft.AzureActiveDirectory";
/**
 * Gets a string representation of a key for an account created by an Azure Active Directory account provider.
 * @param key - The key of an account.
 */
function accountKeyToString(key) {
    return JSON.stringify([
        key.providerId,
        key.providerArgs.host,
        key.providerArgs.clientId,
        key.accountId
    ]);
}
exports.accountKeyToString = accountKeyToString;
/**
 * Gets the key for an account created by an Azure Active Directory account provider from a previously produced string.
 * @param s - A string previously produced from an account key.
 */
function stringToAccountKey(s) {
    var array = JSON.parse(s);
    if (!Array.isArray(array)) {
        return null;
    }
    return {
        providerId: array[0],
        providerArgs: {
            host: array[1],
            clientId: array[2]
        },
        accountId: array[3]
    };
}
exports.stringToAccountKey = stringToAccountKey;
/**
 * The default (public) cloud settings.
 */
exports.defaultSettings = {
    host: "https://login.microsoftonline.com",
    clientId: "872cd9fa-d31f-45e0-9eab-6e460a02d1f1",
    signInResourceId: "https://management.core.windows.net/",
    graphResource: {
        id: "https://graph.windows.net/",
        endpoint: "https://graph.windows.net"
    },
    armResource: {
        id: "https://management.core.windows.net/",
        endpoint: "https://management.azure.com"
    }
};
/**
 * The China cloud settings.
 */
exports.chinaSettings = {
    host: "https://login.chinacloudapi.cn",
    clientId: "872cd9fa-d31f-45e0-9eab-6e460a02d1f1",
    signInResourceId: "https://management.core.chinacloudapi.cn/",
    graphResource: {
        id: "https://graph.chinacloudapi.cn/",
        endpoint: "https://graph.chinacloudapi.cn"
    },
    armResource: {
        id: "https://management.core.chinacloudapi.cn/",
        endpoint: "https://management.chinacloudapi.cn"
    }
};
/**
 * Creates an in-memory ADAL token cache.
 */
function createInMemoryTokenCache() {
    /* tslint:disable no-require-imports */
    return new (require("./token-cache/inMemory").default)();
    /* tslint:enable */
}
exports.createInMemoryTokenCache = createInMemoryTokenCache;
/**
 * Creates a JSON-file based ADAL token cache.
 * @param dir - An optional directory from which the file should be located.
 *              If omitted, defaults to the current working directory.
 * @param filename - An optional name of the file to be used.
 *                   If omitted, defaults to ".adalcache".
 * @param noProbe - Optionally indicates if no probing should occur for the file.
 *                  Probing (the default) starts from the specified directory and
 *                  looks through the parent chain of directories for the file,
 *                  and then looks in the user's home directory. If the file is
 *                  not found, it creates the file in the user's home directory.
 */
function createJsonFileTokenCache(dir, filename, noProbe) {
    /* tslint:disable no-require-imports */
    return new (require("./token-cache/jsonFile").default)(dir, filename, noProbe);
    /* tslint:enable */
}
exports.createJsonFileTokenCache = createJsonFileTokenCache;
/**
 * Creates a JSON-file based ADAL token cache that stores tokens in a native keychain.
 * @param filename - An optional name of the file to be used.
 *                   If omitted, defaults to ".adalcache".
 */
function createJsonFileWithNativeTokenCache(filename) {
    /* tslint:disable no-require-imports */
    return new (require("./token-cache/jsonFileWithNative").default)(filename);
    /* tslint:enable */
}
exports.createJsonFileWithNativeTokenCache = createJsonFileWithNativeTokenCache;
/**
 * The default (public) cloud authorization code grant flow settings.
 */
exports.defaultAuthorizationCodeGrantFlowSettings = _.extend({}, exports.defaultSettings, {
    redirectUri: "urn:ietf:wg:oauth:2.0:oob"
});
/**
 * The China cloud authorization code grant flow settings.
 */
exports.chinaAuthorizationCodeGrantFlowSettings = _.extend({}, exports.chinaSettings, {
    redirectUri: "urn:ietf:wg:oauth:2.0:oob"
});
/**
 * Creates a new instance of the Azure Active Directory account provider that uses the authorization code grant flow.
 * @param getAuthorizationCode - A function that provides the interactive aspect of the flow.
 * @param settings - Optional settings for the flow.
 * @param tokenCache - An optional token cache to use.
 *                     If omitted, defaults to an in-memory token cache.
 */
function createAuthorizationCodeGrantFlowProvider(getAuthorizationCode, settings, tokenCache) {
    /* tslint:disable no-require-imports */
    return new (require("./flow/authorizationCodeGrant").default)(getAuthorizationCode, settings, tokenCache);
    /* tslint:enable */
}
exports.createAuthorizationCodeGrantFlowProvider = createAuthorizationCodeGrantFlowProvider;
